New API: Add Key and Role classes

MVrachev opened the pull request:
Addresses, but doesn't fix: #1139

**Description of the changes being introduced by the pull request**:

  In the top level metadata classes, there are complex attributes such as
  `meta` in Targets and Snapshot, `key` and `roles` in Root etc.
  We want to represent those complex attributes with a class to allow
  easier verification and support for metadata with unrecognized fields.
  For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

The changes in this pr include:
- addition of a new `Key` class  integrated into Root
- addition of a new `Roles` class integrated into 
- added support for unrecognized fields
- added tests for unrecognized fields in `Key` and `Roles`

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
Add Key class and integrate it into Root

In the top level metadata classes, there are complex attributes such as
"meta" in Targets and Snapshot, "key" and "roles" in Root etc.
We want to represent those complex attributes with a class to allow
easier verification and support for metadata with unrecognized fields.
For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add a Role class and integrate it into Root

In the top level metadata classes, there are complex attributes such as
"meta" in Targets and Snapshot, "key" and "roles" in Root etc.
We want to represent those complex attributes with a class to allow
easier verification and support for metadata with unrecognized fields.
For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
There is an interesting reason for the CI [failure](https://github.com/theupdateframework/tuf/pull/1360/checks?check_run_id=2401922288): 
`/home/runner/work/tuf/tuf/tuf/api/metadata.py:532:4: R0914: Too many local variables (18/15) (too-many-locals)`

I used many local variables to make it clear what their purpose is.
Any suggestions? Should I disable this warning with an inline comment or?

jku commented on tuf/api/metadata.py:
```diff
@@ -481,10 +522,14 @@ def from_dict(cls, root_dict: Mapping[str, Any]) -> "Root":
     def to_dict(self) -> Dict[str, Any]:
         """Returns the dict representation of self. """
         root_dict = self._common_fields_to_dict()
+        keys = {}
+        for keyid, key in self.keys.items():
+            keys[keyid] = key.to_dict()
```
Do we have a stance on comprehension? These would be the places to use it... This would translate to I think
`keys = {keyid: key.to_dict() for (keyid, key) in self.keys.items()}`

That might still fit the line length :)

MVrachev replied:
The Google-style guideliness [here](https://google.github.io/styleguide/pyguide.html#27-comprehensions--generator-expressions) says that:
`Okay to use for simple cases.`
This is simple enough and I will try it.

jku commented on tuf/api/metadata.py:
```diff
@@ -407,6 +407,49 @@ def bump_version(self) -> None:
+
+        return res_dict
+
+    def get_public_val(self):
+        """Helper function to access the public value of the key."""
+        return self.keyval["public"]
```
I wonder if we should add these helpers only when we know there are users for them?

For example we won't need this in the client as long as the API provides a verify_with_threshold() function...

MVrachev replied:
Yes, you are right.
I thought it would be useful, but I am not sure about the use case.
Will remove it.

jku commented on tuf/api/metadata.py:
```diff
@@ -450,6 +450,41 @@ def get_public_val(self):
+
+    def __init__(
+        self,
+        keyids: set,
```
`Set[str]` maybe?

That said the input you give when you call this constructor isn't a set I think.

I'm not sure how to handle this right now...

jku commented on tuf/api/metadata.py:
```diff
@@ -513,6 +544,13 @@ def from_dict(cls, root_dict: Mapping[str, Any]) -> "Root":
+            threshold = role_dict.pop("threshold")
+            # All fields left in the role_dict are unrecognized.
+            unrecognized_role_fields = role_dict
+            roles[role_str] = Role(keyids, threshold, unrecognized_role_fields)
```
keyids here probably isn't a set but a list (and Role() hopes it is a set)?

jku commented on tuf/api/metadata.py:
```diff
@@ -525,12 +563,15 @@ def to_dict(self) -> Dict[str, Any]:
         keys = {}
         for keyid, key in self.keys.items():
             keys[keyid] = key.to_dict()
+        roles = {}
+        for role_str, role in self.roles.items():
+            roles[role_str] = role.to_dict()
```
_maybe_ this would still be readable with dict comprehension (I may be stretching it now)?

MVrachev replied:
I tried it and it doesn't look good, because it doesn't fit in the line.

jku commented on tuf/api/metadata.py:
```diff
@@ -397,6 +407,84 @@ def bump_version(self) -> None:
+            **self.unrecognized_fields,
+        }
+
+        return res_dict
```
nit: could just return the dict without a local variable

mnm678 commented:
> There is an interesting reason for the CI [failure](https://github.com/theupdateframework/tuf/pull/1360/checks?check_run_id=2401922288):
> `/home/runner/work/tuf/tuf/tuf/api/metadata.py:532:4: R0914: Too many local variables (18/15) (too-many-locals)`
> 
> I used many local variables to make it clear what their purpose is.
> Any suggestions? Should I disable this warning with an inline comment or?

This is interesting. I think local variables generally make code more readable, so I'm in favor of disabling this warning.

jku commented:
> I used many local variables to make it clear what their purpose is.
> Any suggestions? Should I disable this warning with an inline comment or?

I don't think the function looks horrible but ... you could make `Role.from_dict()` and `Key.from_dict()` -- this would fit the existing style and would probably remove half the variables.

jku commented on tuf/api/metadata.py:
```diff
@@ -407,6 +407,49 @@ def bump_version(self) -> None:
+    ) -> None:
+        self.keytype = keytype
+        self.scheme = scheme
+        self.keyval = keyval
```
I'm fine with storing keyval as is, but should we validate it here (basically check that there is a `keyval["public"]`)? Or are you planning to do all that later?

jku reviewed (commented):
Thanks Martin. I have one potential bug (the keyids type issue which I'm not yet sure how to handle), the rest are just nits/suggestions/questions

jku commented:
There are currently no Key/Role specific tests: We probably should have at least very basic ones if we now have objects?

One thing that might be nice to check for is invalid data -- what does the failure look like if the metadata is not valid (say a missing required field). I don't think this needs to be exhaustively tested but would be nice to see at least an example: I assume with this code we mostly get KeyErrors?

Commit:
Add from_dict methods in Key and Role classes

The from_dict() method simplifies the object creation.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Make keyids in Role a set

From the specification:
"Clients MUST ensure that for any KEYID represented in this key list
and in other files, only one unique key has that KEYID."

The “only one unique key has that KEYID” is a requirement which can’t
be achieved if two keyids are the same.
So, in order to mandate that requirement it makes sense to use a set
which will guarantee us the keyid’s uniqueness.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Check the format of keyval in Root

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
New API: Add Key/Role specific tests

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku commented on tests/test_api.py:
```diff
@@ -365,6 +424,13 @@ def test_support_for_unrecognized_fields(self):
             # Test that the metadata classes store unrecognized fields when
             # initializing and passes them when casting the instance to a dict.
 
+            # Add unrecognized fields to all metadata sub (helper) classes.
+            if metadata == "root":
+                for keyid in dict1["signed"]["keys"].keys():
+                    dict1["signed"]["keys"][keyid]["d"] = "c"
+                for role_str in dict1["signed"]["roles"].keys():
+                    dict1["signed"]["roles"][role_str]["e"] = "g"
```
this looks like it will become hard to maintain when more tests are added but I'm fine with it for now: can you make an issue about doing this in a more structured way (like automate the dictionary poisoning so that ever dictionary is injected, not just ones we list here)?

MVrachev replied:
Yes, you are right. We should automate this.
In a discussion, you showed a prototype automating this process.
Will appreciate it if you propose these changes and we can discuss them, but I don't think this is in the scope of this pr.

jku commented on tuf/api/metadata.py:
```diff
@@ -405,6 +405,96 @@ def bump_version(self) -> None:
+        threshold: int,
+        unrecognized_fields: Optional[Mapping[str, Any]] = None,
+    ) -> None:
+        keyids_set = set(keyids)
+        if len(keyids_set) != len(keyids):
+            raise ValueError(
+                f"keyids should be a set, instead got a list with dublicates!"
+            )
```
typo on duplicate.

More importantly, there's still a mismatch with input and the type hints. If your code expects a set here, then the calling code must make the length check. Alternatively if the check is left here, then the constructor argument type is actually a list. Currently from_dict() already creates a set _without doing the length check_ and then calls this function -- so this error is likely never raised.

jku reviewed (changes requested):
Just one real issue: the keyids uniqueness checks and creation of the set still have a smell around them: I think the error isn't currently raised (when using `from_dict()`).

MVrachev commented:
Updated the pr by:
- changing the type of `keyds` to a list inside the `Role.__init__` 
-  don't make create a set in `Role.from_dict` and left the `Role.__init__` to make that check
- added a unit test case for creating a `Role` instance with duplicates in `keyid`.

MVrachev commented:
Rebased after #1363 was merged.

jku commented on tuf/api/metadata.py:
```diff
@@ -490,17 +577,17 @@ def add_key(
-        if keyid not in self.roles[role]["keyids"]:
-            self.roles[role]["keyids"].append(keyid)
+        if keyid not in self.roles[role].keyids:
+            self.roles[role].keyids.add(keyid)
```
only realized this now but the check here is not needed with a set

jku reviewed (approved):
I think this looks good. I did notice one unneeded "if" that I had missed earlier (adding duplicates to a set is safe).

Commit:
keyid: verify adding an existing key is ignored

Verify that adding an already existing key to keyid for a particular
role in Root won't create duplicate key.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>
